### ISSUE TITLE
Z fighting when applications render using GLSL

### ISSUE THREAD

**Issue by kintel** [floriankirsch/OpenCSG#8]
OpenCSG transforms vertices (and hence produces depth values) using the legacy OpenGL fixed pipeline.
On common GPUs (Tested modern Intel iGPU and nVidia), the fixed function pipeline produces slightly different vertex positions than the programmable pipeline. Since OpenCSG relies on applications rendering final images using a GL_EQUAL depth test, this causes z buffer tearing for applications rendering using a programmable pipeline.

For more info, see https://github.com/openscad/openscad/issues/4595

**Comment by floriankirsch** [comment]
A valid bug. But in OpenCSG, just switching to a different of pipeline will not help, since this would introduce the same problem for applications using the fixed pipeline.

Probably this means that there must be some kind of hook to control how OpenCSG transforms the vertices. That hook would optionally activate some programmable shader. 

Question is whether one hook is sufficient, and if this must be fully controlable from the calling application.

**Comment by kintel** [comment]
Yeah, there are still, but hopefully very few, apps who still use legacy OpenGL. A compile-time flag might be the easiest option, but that will cause challenges for packaging, but will work for embedding.
An alternative to a hook could be to add an initialization function to OpenCSG. That could also help solve the GLEW issue (https://github.com/floriankirsch/OpenCSG/issues/10).

**Comment by floriankirsch** [comment]
I have played with this, using OpenSCAD 2021.01 on Ubuntu 22, and using some NVidia 710GT. So far I have ported the SCS shaders to GLSL.

With the prioprietary NVidia drivers, results for 
  1. gl_Position = gl_ModelViewProjectionMatrix * gl_Vertex;
  2. gl_Position = ftransform();
  3. fixed function
all differ, i.e., you get the depth artifacts if there is a mismatch.

With Nouveau, 1. and 2. appear to have the same results, but 2 and 3 differ. 

My original hope was that 2. and 3. would be a same, and 1. was different, but I did not find this result. Given that, maybe ftransform() is just a useless complication!?

Yet another finding was that with Nouveau, even the existing ARB shaders and the fixed function pipeline differs. That can be even seen in the opencsgexample application.

Currently I believe that is must be possible to switch between option 1, 2, or 3 at runtime.

**Comment by kintel** [comment]
@floriankirsch Awesome! Thanks for looking into this!

**Comment by floriankirsch** [comment]
With Nouveau, the options for position invariance indeed appear to be just broken.

The aforementioned commit is a lame fix that only works if there are less than 256 primitives.

Furthermore, in the last days, I have committed an API to provide GLSL vertex shaders. If those are used, this should resolve the OpenSCAD issues. I have started already an OpenSCAD patch some days ago (https://github.com/floriankirsch/openscad/commit/224c7b4a6cbe841b2a91d0e4c2218f366205e32b). When I have released OpenCSG 1.8.0, I will finalize that.

**Comment by floriankirsch** [comment]
Released with OpenCSG 1.8.0. Closing here.

**Comment by floriankirsch** [comment]
Unfortunately, the fix does not work on systems where only OpenGL 2.1 is available, for example on MacOSX. See https://github.com/openscad/openscad/pull/5725. The problem is that the GLSL shader currently depends on a function that is only available since OpenGL 3.0.

Reopening.

**Comment by floriankirsch** [comment]
Released a fix with OpenCSG 1.8.1, that allows also OpenGL 2.0 and 2.1 to work with the GLSL code path.

The commit for https://github.com/floriankirsch/OpenCSG/issues/8#issuecomment-2689110837 has been reverted in the new release. As shown in https://github.com/openscad/openscad/pull/5725#issuecomment-2707411676 on MacOS X this change caused other rendering issues. 

I close this issue now. For the Z-artifacts that are unrelated to GLSL, a new issue must be opened - it is just unrelated and not clear how to fix.
